common/utils/include/claragenomics/utils/device_preallocated_allocator.cuh

```diff
@@ -0,0 +1,310 @@
+            block_to_the_left.size_  = 0;
+        }
+        else
+        {
+            block_to_the_left = *std::prev(free_blocks_iter);
+            assert(block_to_the_left.begin_ + block_to_the_left.size_ <= block_start);
+            if (block_to_the_left.begin_ + block_to_the_left.size_ == block_start)
+            {
+                // neighbor is "touching" this block and will be merged, remove it
+                free_blocks_.erase(std::prev(free_blocks_iter));
+            }
+            else
+            {
+                // neighbor won't be merged, create a virtual empty neighbor
+                block_to_the_left.begin_ = block_start;
+                block_to_the_left.size_  = 0;
+            }
+        }
+
+        // * find the left neighbor and remove if it is going to be be merged
+        MemoryBlock block_to_the_right;
+        if (std::end(free_blocks_) == free_blocks_iter)
+        {
+            // no neighbor to the right, create a virtual empty neighbor
+            block_to_the_right.begin_ = block_start + number_of_bytes;
+            block_to_the_right.size_  = 0;
+        }
+        else
+        {
+            block_to_the_right = *free_blocks_iter;
+            assert(block_start + number_of_bytes <= block_to_the_right.begin_);
+            if (block_start + number_of_bytes == block_to_the_right.begin_)
+            {
+                // neighbor is "touching" this block and will be merged, remove it
+                auto iter_past_right_neighbor = std::next(free_blocks_iter);
+                free_blocks_.erase(free_blocks_iter);
+                free_blocks_iter = iter_past_right_neighbor;
+            }
+            else
+            {
+                // neighbor won't be merged, create a virtual empty neighbor
+                block_to_the_right.begin_ = block_start + number_of_bytes;
+                block_to_the_right.size_  = 0;
+            }
+        }
+
+        // create the new free memory block
+        MemoryBlock new_memory_block;
+        // block_to_the_left.begin_ corresponds to begin_ of the newly freed block if the left
+        // neighbor should not be merged, begin_ of the actual left neighbor otherwise
+        new_memory_block.begin_ = block_to_the_left.begin_;
+        // block_to_the_left.size_ and block_to_the_right.size_ have value 0 if left and right neighbors
+        // should not be merged, number of elements in left and right neighbor otherwise
+        new_memory_block.size_ = block_to_the_left.size_ + number_of_bytes + block_to_the_right.size_;
+
+        free_blocks_.insert(free_blocks_iter, new_memory_block);
+
+        return cudaSuccess;
+    }
+
+    /// buffer size
+    const size_t buffer_size_;
+    /// preallocated buffer
+    const std::unique_ptr<char, void (*)(char*)> buffer_ptr_;
+    /// (de)allocation mutex
+    mutable std::mutex memory_operation_mutex_;
+
+    /// list of free block, sorted
+    std::list<MemoryBlock> free_blocks_;
```

Comment:
I did think about that, but decided to go for list because vector would also require reallocation from time to time and list sounded like a better abstraction here.
We are anyway talking about up to a few hundred/thousands allocated arrays at a time so I don't think either option is really bad performance wise.

I do agree about comment and added it

Reply:
Ok.
Just for future reference: It is not just because of the memory allocations, but also really slow linear search on std::list due to bad prefetching. The processor will not be able to prefetch the next elements, because it has to load the element in order to know the next location.